lib/chef/knife/cloud/google_service.rb

```diff
@@ -0,0 +1,491 @@
+
+      # no image project has been provided. We'll first check the user's project for the image.
+      url = image_url_for(project, image)
+      return url unless url.nil?
+
+      # Image not found in user's project. Is there a public project this image might exist in?
+      public_project = public_project_for_image(image)
+      if public_project
+        return image_url_for(public_project, image)
+      end
+
+      # No image in user's project or public project, so it doesn't exist.
+      nil
+    end
+
+    def image_url_for(image_project, image_name)
+      return "projects/#{image_project}/global/images/#{image_name}" if image_exist?(image_project, image_name)
+    end
+
+    def boot_disk_name_for(options)
+      options[:boot_disk_name].nil? ? options[:name] : options[:boot_disk_name]
+    end
+
+    def machine_type_url_for(machine_type)
+      "zones/#{zone}/machineTypes/#{machine_type}"
+    end
+
+    def instance_metadata_for(metadata)
+      return if metadata.nil? || metadata.empty?
+
+      metadata_obj = Google::Apis::ComputeV1::Metadata.new
+      metadata_obj.items = metadata.each_with_object([]) do |(k, v), memo|
+        metadata_item       = Google::Apis::ComputeV1::Metadata::Item.new
+        metadata_item.key   = k
+        metadata_item.value = v
+
+        memo << metadata_item
+      end
+
+      metadata_obj
+    end
+
+    def instance_network_interfaces_for(options)
+      interface = Google::Apis::ComputeV1::NetworkInterface.new
+      interface.network = network_url_for(options[:network])
+      interface.access_configs = instance_access_configs_for(options[:public_ip])
+
+      Array(interface)
+    end
+
+    def instance_access_configs_for(public_ip)
+      return [] if public_ip.nil? || public_ip == "NONE"
+
+      access_config = Google::Apis::ComputeV1::AccessConfig.new
+      access_config.name = "External NAT"
+      access_config.type = "ONE_TO_ONE_NAT"
+      access_config.nat_ip = public_ip if valid_ip_address?(public_ip)
+
+      Array(access_config)
+    end
+
+    def network_url_for(network)
+      "projects/#{project}/global/networks/#{network}"
+    end
+
+    def instance_scheduling_for(options)
+      scheduling = Google::Apis::ComputeV1::Scheduling.new
+      scheduling.automatic_restart = options[:auto_restart].to_s
+      scheduling.on_host_maintenance = migrate_setting_for(options[:auto_migrate])
+
+      scheduling
+    end
+
+    def migrate_setting_for(auto_migrate)
+      auto_migrate ? "MIGRATE" : "TERMINATE"
+    end
+
+    def instance_service_accounts_for(options)
+      return if options[:service_account_scopes].nil? || options[:service_account_scopes].empty?
+
+      service_account = Google::Apis::ComputeV1::ServiceAccount.new
+      service_account.email  = options[:service_account_name]
+      service_account.scopes = options[:service_account_scopes].map { |scope| "https://www.googleapis.com/auth/#{scope}" unless scope.start_with?("https://www.googleapis.com/auth/") }
```

Comment:
It might be good to allow for the same scope aliases that the Cloud SDK (gcloud) supports. You could do something similar with what you did for the image projects.

   ```
  # The help page shows the list of aliases
  gcloud compute instances create --help
  ```

Since the setup instructions for knife-google suggest using gcloud, it's likely the user may use both.